src/main/java/com/scalyr/integrations/kafka/AddEventsClient.java

```diff
@@ -215,47 +221,53 @@ private AddEventsResponse addEventsWithRetry(byte[] addEventsPayload, long uncom
         byte[] decompressedPayload = getDecompressedPayload(addEventsPayload);
         payloadLogger.error("Add events too large payload: {}", new String(decompressedPayload));
       }
-      return PAYLOAD_TOO_LARGE;
+      addEventsResult.complete(PAYLOAD_TOO_LARGE);
+      return;
     }
 
-    AddEventsResponse addEventsResponse = TIMEOUT_ADD_EVENTS_RESPONSE;
-    long delayTimeMs = initialBackoffDelayMs;
-    httpPost.setEntity(new ByteArrayEntity(addEventsPayload));
+    // Retries timed out
+    if (remainingMs(startTimeMs) == 0) {
+      addEventsResult.complete(TIMEOUT_ADD_EVENTS_RESPONSE);
+      return;
+    }
 
-    boolean shouldCallAddEvents = remainingMs(startTimeMs) > 0;
-    while (shouldCallAddEvents) {
-      try (CloseableHttpResponse httpResponse = client.execute(httpPost)) {
-        addEventsResponse = parseAddEventsResponse(httpResponse);
-        log.debug("post http code {}, httpResponse {}", httpResponse.getStatusLine().getStatusCode(), addEventsResponse);
-        // return on success or non-retriable error
+    requestBuilder.setBody(addEventsPayload);
+    final ListenableFuture<Response> addEventsFuture = requestBuilder.execute();
+    addEventsFuture.addListener(() -> {
+      AddEventsResponse addEventsResponse;
+      try {
+        final Response response = addEventsFuture.get();
+        addEventsResponse = parseAddEventsResponse(response);
+        log.debug("post http code {}, httpResponse {}", response.getStatusCode(), addEventsResponse);
         if (addEventsResponse.isSuccess() || !addEventsResponse.isRetriable()) {
-          return addEventsResponse;
+          addEventsResult.complete(addEventsResponse);
+          return;
         }
-      } catch (IOException e) {
-        log.warn("Error calling Scalyr addEvents API", e);
-        addEventsResponse = new AddEventsResponse().setStatus("IOException").setMessage(e.toString());
+      } catch (Exception e) {
+        Throwable rootCauseException = e.getCause() == null ? e : e.getCause();
+        addEventsResponse = new AddEventsResponse().setStatus("addEvents error").setMessage(rootCauseException.toString());
       }
 
-      shouldCallAddEvents = remainingMs(startTimeMs) > delayTimeMs;
-      if (shouldCallAddEvents) {
-        sleep.accept(delayTimeMs);
-        delayTimeMs = delayTimeMs * 2;
+      // error - retry if not timed out and retriable, otherwise, complete with failure
+      log.warn("addEvents failed with {}", addEventsResponse);
+      if (remainingMs(startTimeMs) > retryDelayMs && addEventsResponse.isRetriable()) {
+        sleep.accept(retryDelayMs);
```

Comment:
I'm not that familiar with Java async patterns, but I assume ``Uninterruptibles.sleepUninterruptibly`` plays along nicely with the async code and doesn't block the whole event loop?

Reply:
` Uninterruptibles.sleepUninterruptibly` does block.  I changed this to use a `ScheduledExecutorService`.  See response on line 79 for more details.